### ISSUE TITLE
Berechnung des PV-Anteils im Power-Splitter

### ISSUE THREAD

**Issue by deThommy** [solectrus/power-splitter#125]
### Beschreibung (auf Deutsch oder Englisch)

Hallo Georg,

vorweg: ich habe aktuell keine Lösung für das nachfolgende "Problem", es ist nur eine Beobachtung, die mir auffiel, da aktuell praktisch keine Sonne vorhanden ist.

Mein Hausakku wird in Zeiten günstiger Strompreise aus dem Netz beladen, damit ich in zu teureren Zeiten dann von dem günstiger eingekauften Netz-Strom profitieren kann. So wurde mein Hausakku in der vergangenen Nacht vollständig aus dem Netz geladen. Aktuell bedient sich die Wärmepumpe jetzt natürlich komplett an dem Akku-Netz-Strom. 

Es wird jetzt im Power-Splitter so dargestellt, als wäre das "PV-Energie" (da sie aus dem Hausakku kommt). Erstmal logisch. Aber aufgrund der Netzbeladung des Hausakkus natürlich sachlich richtig.

Bekommt man das irgendwie differenziert? Theoretisch "weiß" der Hausakku ja, dass in ihm aktuell nur Netzstrom und kein PV-Strom vorhanden ist, da Solectrus mitbekommt, dass der Hausakku aus dem Netz beladen wurde.

Viele Grüße, deThommy.

### Welche Version von SOLECTRUS verwendest du?

0.17.1

### Auf welchem Gerät läuft bei dir SOLECTRUS?

Synology

**Comment by ledermann** [comment]
Ja, das ist ein grundsätzliches Problem. Der Power-Splitter geht davon aus, dass Strom aus dem Akku ursprünglich durch PV entstanden ist. In dem von dir beschriebenen Fall der Akkubeladung aus dem Netz ist das dann **nicht korrekt**.

> Theoretisch "weiß" der Hausakku ja, dass in ihm aktuell nur Netzstrom und kein PV-Strom vorhanden ist, [...]

Ja, aber eben nur **theoretisch**, denn derzeitig wird das eben nicht nachgehalten. Außerdem wird im Speicher meist ein Strommix enthalten sein.

Denkbar wäre aber vielleicht, den PV-Anteil im Akku jederzeit zu berechnen. Wenn das dann bekannt wäre, könnte bei der Berechnung des PV-Anteils der Verbraucher korrigiert werden.

Ich mache mir da mal Gedanken zu. Falls jemand Ideen hat - nur her damit! :)

**Comment by deThommy** [comment]
Ich habe leider gar keine Progammier-Erfahrung oder mit den hier zugrundeliegenden Datenbanken. 

In der Darstellung von Solectrus sieht die Akkubeladung durch Netz-Strom korrekt dargestellt aus: sprich 100% Netzbezug, der bspw. zu 90% in den Akku fließt und 10% direkt verbraucht wird.

Kannst du dem aktuell fließenden Strom vielleicht eine Art "Tag" verpassen, sodass der Akku berechnen kann, zu wie viel Prozent PV und Netz-Strom vorhanden ist? Dann könnte auf Basis dessen bspw. berechnet werden, dass der Akku aktuell zu x% PV enthält, was wiederum für die Berechnung im Power-Splitter verwendet werden kann?

Vielleicht meintest du das so auch schon mit deinem Vorschlag, fällt mir gerade auf!?

**Comment by ledermann** [comment]
Ja, so in etwa dachte ich mir das. Die Berechnung ist aber nicht trivial. Ich habe da ein wenig mit experimentiert, eine schnelle Lösung ist da aber nicht absehbar. 

Erforderlich wäre das übrigens auch bei der Notbeladung, die im Winter öfter mal durchgeführt wird, wenngleich es sich da ja nur um 1-2 kWh handelt. Die Auswirkungen der Notbeladungen auf den Power-Splitter sind also sehr gering.

**Comment by Robbe64** [comment]
Ich schließe mich hier mal an.
Allerdings erstmal mit einer etwas einfacheren Problemstellung wenn ich euch richtig verstehe.
Und zwar könnte man ja im ersten Schritt bei der Akkubeladung, nach dem gleichen Prinzip wie beim Auto Laden oder dem Hausverbrauch, einen Netz und PV Anteil berechnen und darstellen. Dann könnte man in der Statistik schonmal sehen, wie viel Strom ich z.B. im Jahr aus dem Netz in den Akku geladen wurde.

Ob der Stom der aus dem Akku kommt jetzt Netz oder PV Stom ist wäre dann der zweite Schritt, der dann hoffentlch darauf aufbauen könnte.

**Comment by ledermann** [comment]
@Robbe64 Ja, für den Akku zunächst einmal zu berechnen, wie viel Netzbezug in die Beladung geflossen ist, ist leicht machbar. Ich habe den PowerSplitter entsprechend erweitert. Eine einfache Darstellung im Dashboard habe ich ebenfalls vorgenommen. Hier ein [Beispiel](https://demo.solectrus.de/battery_power/2024): 

![Image](https://github.com/user-attachments/assets/2490a5dd-44f7-42d9-b5f5-10546f54721f)

Bei mir kommt der Netzbezug des Akkus übrigens ausschließlich über die Notladungen - und ein kleiner Rest über die Trägheit des Systems. Das werde ich mir nun mal eine Weile ansehen.

Allerdings ist unklar, wie es damit weitergehen soll. Denn letztendlich wird der Akku aus dem Netz beladen, um zeitversetzt günstigen Strom für den Verbrauch zu haben. Das hießt, der Netzbezug des Akkus muss auf die Verbraucher verteilt werden. Und da wird es richtig kompliziert. Da habe ich noch keine Idee, wie das mit vertretbarem Aufwand machbar wäre.

Außerdem: So richtig Sinn ergibt das ganze erst, wenn SOLECTRUS [mit dynamischen Stromtarifen umgehen](https://github.com/solectrus/solectrus/issues/2150) kann, die Kosten also korrekt berechnen kann. Das ist aber noch nicht implementiert - nicht zuletzt deshalb, weil ich selbst weiterhin einen Festpreistarif habe.

**Comment by deThommy** [comment]
Wäre es nicht möglich den durchschnittlichen "Strommix" des Akkus (also das, was Du gerade in Deinen Screenshots geändert hast) für die weitere Berechnung im PowerSplitter zu nutzen. 

Ich weiß nicht, wie viel das bringen würde, wäre in dem Szenario bei mir aber erheblich genauer (was die Kosten angeht). 

Hier mal ein Beispiel aus Dezember: den PV-Prozentsatz des E-Autos glaube ich sofort. Bei der Wärmepumpe stimmt das hinten und vorne nicht (das grenzt im Dezember vermutlich an 0%), was sich natürlich dementsprechend auch auf die Kostenberechnung auswirkt. Auch dafür wäre der Preis des "Strommixes" (also Anteil Netzbezug zu x ct/kWh und Anteil PV zu y ct/kWh). Die 39% des PowerSplitters entsprechend hier ja schon fast 220 kWh, wobei hier im Dezember leider nur 101 kWh Ertrag waren -- vielleicht lässt sich auch davon eine Logik ableiten?

![Image](https://github.com/user-attachments/assets/c0d7df57-3ccb-44bc-bbd9-4a00b85c5227)

![Image](https://github.com/user-attachments/assets/24ffef34-8ee3-46db-b2d4-18fc8fa1b5c2)

So stellt sich mir leider gerade die Frage nach dem Nutzen des PowerSplitters und ich würde vermuten, dass noch etliche andere dieses Problem haben müssten (oder bald haben werden, da die Vermarktung von diesen "Smarten HEMS" ja jetzt in meiner Wahrnehmung schon deutlich mehr wird.)

Danke jedenfalls, dass du Dich mit dem Thema beschäftigst!!

**Comment by ledermann** [comment]
@deThommy Danke für dein Feedback! Deinen Screenshots entnehme ich, dass du im Dezember wohl erheblich den Speicher aus dem Netz geladen hast, denn trotz nur 101 kWh Erzeugung sind 298 kWh in den Akku geflossen. Das ist genau der Kern des Problems, denn der PowerSplitter interpretiert derzeitig den Strom, der aus dem Akku entnommen wird, als PV-Strom. Daher wird bei deiner Wärmepumpe ein zu hoher PV-Anteil ausgewiesen, die 38 % stimmen natürlich nicht.

Nochmal zum Hintergrund: Der PowerSplitter ermittelt derzeitig den Anteil des Stroms, der für einen Verbraucher aus dem Netz bezogen wurde (und zwar zum Zeitpunkt des Verbrauchs). Alles andere wird als PV-Strom interpretiert. Dieser Ansatz müsste verändert werden, um Beladungen des Speichers aus dem Netz richtig zu behandeln.

Meine kürzlich vorgenommene Ergänzung (nach dem Vorschlag von @Robbe64) ist ein erster Schritt, weil damit ersichtlich wird, welchen PV-Anteil die Akkubeladung hat. Du kannst das bei dir ausprobieren, indem du den PowerSplitter auf `develop` wechselst, eine Neuberechnung veranlasst (`docker kill ...`) und dann die Tageszusammenfassungen zurücksetzt.

Fraglich ist jetzt, wie es damit weitergeht. Dein ganz oben gemachter Vorschlag, dass der PowerSplitter jederzeitig den PV-Anteil des im Speicher befindlichen Stroms kennen müsste, wäre ein Ansatz. Ich komme wohl nicht daran vorbei, mir dazu weiter Gedanken zu machen ;-)
